Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+{-# OPTIONS_GHC -Wall #-}
+
+module Data.Primitive.PrimArray
+  ( -- * Types
+    PrimArray(..)
+  , MutablePrimArray(..)
+    -- * Allocation
+  , newPrimArray
+    -- * Element Access
+  , readPrimArray
+  , writePrimArray
+  , indexPrimArray
+    -- * Freezing and Thawing
+  , unsafeFreezePrimArray
+  , unsafeThawPrimArray
+    -- * Block Operations
+  , copyPrimArray
+  , copyMutablePrimArray
+  , copyPrimArrayToPtr
+  , copyMutablePrimArrayToPtr
+  , setPrimArray
+    -- * Information
+  , sameMutablePrimArray
+  , getSizeofMutablePrimArray
+  , sizeofPrimArray
+  ) where
+
+import GHC.Prim
+import GHC.Exts (isTrue#,IsList(..))
+import GHC.Int
+import GHC.Ptr
+import Data.Primitive
+import Control.Monad.Primitive
+import Control.Monad.ST
+import Data.Semigroup (Semigroup(..))
+import qualified Data.Semigroup as SG
+import qualified Data.List as L
+import qualified Data.Primitive.ByteArray as PB
+import qualified Data.Primitive.Types as PT
+
+-- | Primitive arrays
+data PrimArray a = PrimArray ByteArray#
+
+-- | Mutable primitive arrays associated with a primitive state token
+data MutablePrimArray s a = MutablePrimArray (MutableByteArray# s)
+
+instance (Eq a, Prim a) => Eq (PrimArray a) where
+  a1 == a2 = sizeofPrimArray a1 == sizeofPrimArray a2 && loop (sizeofPrimArray a1 - 1)
+    where 
+    loop !i | i < 0 = True
+            | otherwise = indexPrimArray a1 i == indexPrimArray a2 i && loop (i-1)
+
+instance Prim a => IsList (PrimArray a) where
+  type Item (PrimArray a) = a
+  fromList xs = primArrayFromList (L.length xs) xs
+  fromListN = primArrayFromList
+  toList = primArrayToList
+
+instance (Prim a, Show a) => Show (PrimArray a) where
+  showsPrec p = showsPrec p . primArrayToList
+
+primArrayFromList :: forall a. Prim a => Int -> [a] -> PrimArray a
+primArrayFromList len vs = runST run where
+  run :: forall s. ST s (PrimArray a)
+  run = do
+    arr <- newPrimArray len
+    let go :: [a] -> Int -> ST s ()
+        go !xs !ix = case xs of
+          [] -> return ()
+          a : as -> do
+            writePrimArray arr ix a
+            go as (ix + 1)
+    go vs 0
+    unsafeFreezePrimArray arr
+
+primArrayToList :: forall a. Prim a => PrimArray a -> [a]
+primArrayToList arr = go 0 where
+  !len = sizeofPrimArray arr
+  go :: Int -> [a]
+  go !ix = if ix < len
+    then indexPrimArray arr ix : go (ix + 1)
+    else []
```

Comment:
I suggest adding folds, and then copying the default definition of `toList` to implement this; that'll get you list fusion, and of course we want folds anyway.